File: market_data_pipeline/src/connectors/polymarket.py
"""Polymarket CLOB + Gamma API connector.

Polymarket exposes two public APIs (no auth required):
  - **Gamma API** (gamma-api.polymarket.com) — event/market metadata, categories,
    rich descriptions, outcome prices.
  - **CLOB API** (clob.polymarket.com) — order-book data, price history, spreads.

This connector fetches active prediction markets and price histories, returning
data compatible with the pipeline's adapter contract.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import date, datetime, timezone
from typing import Any

from .base import RateLimiter, ResponseCache, ThrottledClient

logger = logging.getLogger("market_data_pipeline.connectors.polymarket")

GAMMA_BASE = "https://gamma-api.polymarket.com"
CLOB_BASE = "https://clob.polymarket.com"


@dataclass
class PolymarketMarket:
    condition_id: str
    question: str
    category: str
    yes_price: float
    no_price: float
    spread: float
    volume_24h: float
    total_volume: float
    liquidity: float
    end_date: str
    active: bool


@dataclass
class PolymarketPricePoint:
    timestamp: int
    price: float


class PolymarketConnector:
    """Public Polymarket API connector with rate limiting and caching."""

    def __init__(
        self,
        rate: float = 2.0,
        cache_ttl_hours: int = 1,
    ) -> None:
        self._client = ThrottledClient(rate=rate, timeout=10.0)
        self._cache = ResponseCache()
        self._cache_ttl = cache_ttl_hours

    def fetch_markets(
        self,
        limit: int = 100,
        active: bool = True,
        category: str | None = None,
    ) -> list[PolymarketMarket]:
        cache_key = ResponseCache.make_key(
            "polymarket", "markets", f"active={active}&limit={limit}", {}
        )
        cached = self._cache.get(cache_key, ttl_hours=self._cache_ttl)
        if cached is not None:
            return self._parse_markets(cached, category)

        params: dict[str, Any] = {
            "active": str(active).lower(),
            "closed": "false",
            "limit": str(limit),
        }
        resp = self._client.get(f"{GAMMA_BASE}/markets", params=params)
        data = resp.json()
        self._cache.set(cache_key, data)
        return self._parse_markets(data, category)

    def fetch_events(
        self,
        limit: int = 50,
        active: bool = True,
    ) -> list[dict[str, Any]]:
        cache_key = ResponseCache.make_key(
            "polymarket", "events", f"active={active}&limit={limit}", {}
        )
        cached = self._cache.get(cache_key, ttl_hours=self._cache_ttl)
        if cached is not None:
            return cached if isinstance(cached, list) else []

        params: dict[str, Any] = {
            "active": str(active).lower(),
            "closed": "false",
            "limit": str(limit),
        }
        resp = self._client.get(f"{GAMMA_BASE}/events", params=params)
        data = resp.json()
        self._cache.set(cache_key, data)
        return data if isinstance(data, list) else []

    def fetch_price_history(
        self,
        token_id: str,
        interval: str = "1d",
        fidelity: int = 60,
    ) -> list[PolymarketPricePoint]:
        cache_key = ResponseCache.make_key(
            "polymarket", "price_history", token_id, {"interval": interval}
        )
        cached = self._cache.get(cache_key, ttl_hours=self._cache_ttl)
        if cached is not None:
            return self._parse_history(cached)

        params: dict[str, Any] = {
            "market": token_id,
            "interval": interval,
            "fidelity": str(fidelity),
        }
        resp = self._client.get(f"{CLOB_BASE}/prices-history", params=params)
        data = resp.json()
        self._cache.set(cache_key, data)
        return self._parse_history(data)

    def _parse_markets(
        self, data: Any, category: str | None = None
    ) -> list[PolymarketMarket]:
        if not isinstance(data, list):
            return []
        markets: list[PolymarketMarket] = []
        for item in data:
            try:
                cat = item.get("category", "Other")
                if category and cat.lower() != category.lower():
                    continue
                prices = item.get("outcomePrices", "[]")
                if isinstance(prices, str):
                    import json as _json
                    prices = _json.loads(prices)
                yes_price = float(prices[0]) if len(prices) > 0 else 0.5
                no_price = float(prices[1]) if len(prices) > 1 else 1.0 - yes_price

                markets.append(
                    PolymarketMarket(
                        condition_id=item.get("conditionId", item.get("id", "")),
                        question=item.get("question", ""),
                        category=cat,
                        yes_price=yes_price,
                        no_price=no_price,
                        spread=float(item.get("spread", 0.02)),
                        volume_24h=float(item.get("volume24hr", 0)),
                        total_volume=float(item.get("volume", 0)),
                        liquidity=float(item.get("liquidity", 0)),
                        end_date=str(item.get("endDate", ""))[:10],
                        active=bool(item.get("active", True)),
                    )
                )
            except (ValueError, KeyError, IndexError) as exc:
                logger.debug("Skipping malformed Polymarket market: %s", exc)
        return markets

    def _parse_history(self, data: Any) -> list[PolymarketPricePoint]:
        history = data.get("history", []) if isinstance(data, dict) else []
        points: list[PolymarketPricePoint] = []
        for item in history:
            try:
                points.append(
                    PolymarketPricePoint(
                        timestamp=int(item.get("t", 0)),
                        price=float(item.get("p", 0)),
                    )
                )
            except (ValueError, KeyError):
                continue
        return points

File: src/lib/server/polymarket.ts
/**
 * Server-only Polymarket API client (CLOB + Gamma).
 *
 * Both APIs are public — no auth key required. Activated unconditionally
 * (unlike FRED which gates on FRED_API_KEY). The route handlers try live
 * first and fall back to SIM on network failure, so this is safe to call
 * in any environment.
 *
 * In-memory TTL cache prevents hammering the API on rapid page navigations.
 */

import { fetchWithProxyFallback } from "@/lib/server/fetchProxy";
import type { PolyMarket, PolyEvent, PolyPricePoint, PolyCategory } from "@/data/polymarket";

const GAMMA_BASE = "https://gamma-api.polymarket.com";
const CLOB_BASE = "https://clob.polymarket.com";
const TIMEOUT_MS = 8000;
const CACHE_TTL_MS = 120_000; // 2 minutes

type CacheEntry = { at: number; data: unknown };
const cache = new Map<string, CacheEntry>();

function cached<T>(key: string): T | undefined {
  const e = cache.get(key);
  if (e && Date.now() - e.at < CACHE_TTL_MS) return e.data as T;
  return undefined;
}

function store(key: string, data: unknown) {
  cache.set(key, { at: Date.now(), data });
}

function parseCategory(raw: string | undefined): PolyCategory {
  const map: Record<string, PolyCategory> = {
    politics: "Politics", crypto: "Crypto", economics: "Economics",
    sports: "Sports", science: "Science", culture: "Culture",
    "pop culture": "Culture", tech: "Tech", technology: "Tech",
    climate: "Climate", "climate & weather": "Climate",
  };
  return map[(raw ?? "").toLowerCase()] ?? "Culture";
}

function gammaToMarket(item: any): PolyMarket | null {
  try {
    let prices = item.outcomePrices ?? "[]";
    if (typeof prices === "string") prices = JSON.parse(prices);
    const yesPrice = Number(prices[0] ?? 0.5);
    const noPrice = Number(prices[1] ?? 1 - yesPrice);
    const spread = Math.abs(1 - yesPrice - noPrice);

    let clobIds = item.clobTokenIds ?? "[]";
    if (typeof clobIds === "string") clobIds = JSON.parse(clobIds);

    return {
      id: item.conditionId ?? item.id ?? "",
      question: item.question ?? "",
      category: parseCategory(item.groupItemTitle ?? item.category),
      yesPrice: Number(yesPrice.toFixed(2)),
      noPrice: Number(noPrice.toFixed(2)),
      spread: Number(spread.toFixed(3)),
      volume24h: Number(item.volume24hr ?? 0),
      totalVolume: Number(item.volume ?? 0),
      liquidity: Number(item.liquidity ?? 0),
      chg24h: 0,
      endDate: (item.endDate ?? "").slice(0, 10),
      spark: [],
      active: Boolean(item.active),
    };
  } catch {
    return null;
  }
}

export async function fetchLiveMarkets(opts: {
  limit?: number;
  category?: string;
}): Promise<PolyMarket[]> {
  const limit = opts.limit ?? 100;
  const cacheKey = `poly-markets-${limit}`;
  const hit = cached<PolyMarket[]>(cacheKey);
  if (hit) {
    return opts.category ? hit.filter((m) => m.category === opts.category) : hit;
  }

  const params = new URLSearchParams({
    active: "true",
    closed: "false",
    limit: String(limit),
    order: "volume24hr",
    ascending: "false",
  });
  const res = await fetchWithProxyFallback(
    `${GAMMA_BASE}/markets?${params}`,
    { signal: AbortSignal.timeout(TIMEOUT_MS) }
  );
  if (!res.ok) throw new Error(`Gamma API ${res.status}`);
  const raw: any[] = await res.json();
  const markets = raw.map(gammaToMarket).filter((m): m is PolyMarket => m !== null);
  store(cacheKey, markets);
  return opts.category ? markets.filter((m) => m.category === opts.category) : markets;
}

export async function fetchLiveEvents(limit = 20): Promise<PolyEvent[]> {
  const cacheKey = `poly-events-${limit}`;
  const hit = cached<PolyEvent[]>(cacheKey);
  if (hit) return hit;

  const params = new URLSearchParams({
    active: "true",
    closed: "false",
    limit: String(limit),
  });
  const res = await fetchWithProxyFallback(
    `${GAMMA_BASE}/events?${params}`,
    { signal: AbortSignal.timeout(TIMEOUT_MS) }
  );
  if (!res.ok) throw new Error(`Gamma events API ${res.status}`);
  const raw: any[] = await res.json();

  const events: PolyEvent[] = raw
    .filter((e) => Array.isArray(e.markets) && e.markets.length > 0)
    .map((e) => {
      const eventMarkets = e.markets.map(gammaToMarket).filter((m: PolyMarket | null): m is PolyMarket => m !== null);
      return {
        id: e.id ?? "",
        title: e.title ?? "",
        category: parseCategory(e.category),
        markets: eventMarkets,
        totalVolume: eventMarkets.reduce((s: number, m: PolyMarket) => s + m.totalVolume, 0),
      };
    })
    .filter((e) => e.markets.length > 0);

  store(cacheKey, events);
  return events;
}

export async function fetchLivePriceHistory(
  tokenId: string,
  days = 90
): Promise<PolyPricePoint[]> {
  const cacheKey = `poly-hist-${tokenId}-${days}`;
  const hit = cached<PolyPricePoint[]>(cacheKey);
  if (hit) return hit;

  const params = new URLSearchParams({
    market: tokenId,
    interval: "1d",
    fidelity: "60",
  });
  const res = await fetchWithProxyFallback(
    `${CLOB_BASE}/prices-history?${params}`,
    { signal: AbortSignal.timeout(TIMEOUT_MS) }
  );
  if (!res.ok) throw new Error(`CLOB price-history ${res.status}`);
  const json = await res.json();
  const history: { t: number; p: number }[] = json.history ?? [];

  const points: PolyPricePoint[] = history.map((h) => ({
    date: new Date(h.t * 1000).toISOString().slice(0, 10),
    price: Number(h.p.toFixed(3)),
  }));

  store(cacheKey, points);
  return points;
}
